File: sdk/python/cryptominds/escrow.py
"""
托管客户端

链上资金安全保障，11 态状态机管理。
"""

from typing import Dict, Optional
from enum import Enum
import requests


class EscrowState(Enum):
    """托管状态"""
    CREATED = "created"
    FUNDED = "funded"
    DELIVERED = "delivered"
    CONFIRMED = "confirmed"
    DISPUTED = "disputed"
    ARBITRATING = "arbitrating"
    RELEASED = "released"
    REFUNDED = "refunded"
    SLASHED = "slashed"
    CANCELLED = "cancelled"
    EXPIRED = "expired"


class EscrowClient:
    """托管客户端"""

    def __init__(self, base_url: str = "https://api.cryptominds.ai"):
        self.base_url = base_url

    def create(
        self,
        buyer: str,
        seller: str,
        amount: float,
        token: str = "BNB",
        timeout: int = 86400,
        metadata: Optional[Dict] = None
    ) -> Dict:
        """
        创建托管

        Args:
            buyer: 买家地址
            seller: 卖家地址
            amount: 托管金额
            token: 代币类型 (BNB/BUSD/USDT)
            timeout: 超时时间（秒）
            metadata: 附加数据

        Returns:
            {"escrow_id": "0x...", "state": "created", ...}
        """
        payload = {
            "buyer": buyer,
            "seller": seller,
            "amount": amount,
            "token": token,
            "timeout": timeout,
        }
        if metadata:
            payload["metadata"] = metadata

        resp = requests.post(f"{self.base_url}/api/v1/escrow/create", json=payload)
        resp.raise_for_status()
        return resp.json()

    def get(self, escrow_id: str) -> Dict:
        """查询托管状态"""
        resp = requests.get(f"{self.base_url}/api/v1/escrow/{escrow_id}")
        resp.raise_for_status()
        return resp.json()

    def fund(self, escrow_id: str, tx_hash: str) -> Dict:
        """确认资金已托管（买家调用）"""
        resp = requests.post(
            f"{self.base_url}/api/v1/escrow/{escrow_id}/fund",
            json={"tx_hash": tx_hash}
        )
        resp.raise_for_status()
        return resp.json()

    def deliver(self, escrow_id: str, proof: Dict) -> Dict:
        """提交交付证明（卖家调用）"""
        resp = requests.post(
            f"{self.base_url}/api/v1/escrow/{escrow_id}/deliver",
            json={"proof": proof}
        )
        resp.raise_for_status()
        return resp.json()

    def confirm(self, escrow_id: str) -> Dict:
        """确认交付，释放资金（买家调用）"""
        resp = requests.post(f"{self.base_url}/api/v1/escrow/{escrow_id}/release")
        resp.raise_for_status()
        return resp.json()

    def refund(self, escrow_id: str) -> Dict:
        """申请退款（买家调用，需满足条件）"""
        resp = requests.post(f"{self.base_url}/api/v1/escrow/{escrow_id}/refund")
        resp.raise_for_status()
        return resp.json()

File: sdk/javascript/escrow.js
/**
 * 托管客户端
 * 链上资金安全保障，11 态状态机管理
 * @module escrow
 */

const axios = require('axios');

/**
 * 托管状态枚举
 */
const EscrowState = {
  CREATED: 'created',
  FUNDED: 'funded',
  DELIVERED: 'delivered',
  CONFIRMED: 'confirmed',
  DISPUTED: 'disputed',
  ARBITRATING: 'arbitrating',
  RELEASED: 'released',
  REFUNDED: 'refunded',
  SLASHED: 'slashed',
  CANCELLED: 'cancelled',
  EXPIRED: 'expired',
};

class EscrowClient {
  /**
   * @param {string} baseUrl - API 基础 URL
   */
  constructor(baseUrl = 'https://api.cryptominds.ai') {
    this.baseUrl = baseUrl;
    this.client = axios.create({ baseURL: baseUrl });
  }

  /**
   * 创建托管
   * @param {Object} options - 托管选项
   * @param {string} options.buyer - 买家地址
   * @param {string} options.seller - 卖家地址
   * @param {number} options.amount - 托管金额
   * @param {string} [options.token='BNB'] - 代币类型
   * @param {number} [options.timeout=86400] - 超时时间（秒）
   * @param {Object} [options.metadata] - 附加数据
   * @returns {Promise<Object>} 托管信息
   */
  async create({ buyer, seller, amount, token = 'BNB', timeout = 86400, metadata }) {
    const payload = { buyer, seller, amount, token, timeout };
    if (metadata) {
      payload.metadata = metadata;
    }
    const resp = await this.client.post('/api/v1/escrow/create', payload);
    return resp.data;
  }

  /**
   * 查询托管状态
   * @param {string} escrowId - 托管 ID
   * @returns {Promise<Object>} 托管信息
   */
  async get(escrowId) {
    const resp = await this.client.get(`/api/v1/escrow/${escrowId}`);
    return resp.data;
  }

  /**
   * 确认资金已托管（买家调用）
   * @param {string} escrowId - 托管 ID
   * @param {string} txHash - 交易哈希
   * @returns {Promise<Object>} 更新后的托管信息
   */
  async fund(escrowId, txHash) {
    const resp = await this.client.post(`/api/v1/escrow/${escrowId}/fund`, {
      tx_hash: txHash,
    });
    return resp.data;
  }

  /**
   * 提交交付证明（卖家调用）
   * @param {string} escrowId - 托管 ID
   * @param {Object} proof - 交付证明
   * @returns {Promise<Object>} 更新后的托管信息
   */
  async deliver(escrowId, proof) {
    const resp = await this.client.post(`/api/v1/escrow/${escrowId}/deliver`, {
      proof,
    });
    return resp.data;
  }

  /**
   * 确认交付，释放资金（买家调用）
   * @param {string} escrowId - 托管 ID
   * @returns {Promise<Object>} 更新后的托管信息
   */
  async confirm(escrowId) {
    const resp = await this.client.post(`/api/v1/escrow/${escrowId}/release`);
    return resp.data;
  }

  /**
   * 申请退款（买家调用，需满足条件）
   * @param {string} escrowId - 托管 ID
   * @returns {Promise<Object>} 更新后的托管信息
   */
  async refund(escrowId) {
    const resp = await this.client.post(`/api/v1/escrow/${escrowId}/refund`);
    return resp.data;
  }
}

module.exports = { EscrowClient, EscrowState };